### PR TITLE
Replace overlay2 mount reexec with in-proc impl

### PR DESCRIPTION
**Pull request by cpuguy83**
Building off insights from the great work Cory Snider has been doing, this replaces a reexec with a much lower overhead implementation which performs the `Chddir` in a new goroutine that is locked to a specific thread with CLONE_FS unshared.
The thread is thrown away afterwards and the Chdir does effectively the same thing as what the reexec was being used for.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

### CODE REVIEW

**Comment by thaJeztah** [comment]
WOW; that looks _so much cleaner!_ 😍 

@corhere ptal

File: daemon/graphdriver/overlay2/mount.go
Comment (corhere) [inline daemon/graphdriver/overlay2/mount.go:13]: Nit: the channel does not need to be buffered. Synchronized rendezvous between the two goroutines is the desired behaviour here.
```suggestion
	chErr := make(chan error)
```

**Review by corhere — APPROVED** [review]
Welp, you beat me to opening a PR. 😜 I have a nearly identical commit on a local branch...
```
commit f2761b5ccd000ee760dc1d262a1f5e7df6e49767
Author: Cory Snider <csnider@mirantis.com>
Date:   Mon Oct 3 18:39:59 2022 -0400

    overlay2: refactor mountFrom w/o reexec

    Signed-off-by: Cory Snider <csnider@mirantis.com>
```

File: daemon/graphdriver/overlay2/mount.go
Comment (cpuguy83) [reply daemon/graphdriver/overlay2/mount.go:13]: Agree it's not strictly needed since there is no case where we would leak the goroutine, however we don't need the goroutine to stay running once it's done mounting.

**Review by fuweid — APPROVED** [review]
LGTM(nb)

**Review by thaJeztah — APPROVED** [review]
LGTM 👍

**Comment by fuweid** [comment]
@thaJeztah is it going to do release? The go has issue about thread unlock https://github.com/golang/go/issues/56243. Maybe we should hold it

**Comment by corhere** [comment]
The golang issue is labeled `release-blocker` so we can count on it being fixed in time for Go 1.20 (around Feb/March) and most likely will be backported to 1.19/1.18 sooner. I say we keep this change in master; we can always revert it on a release branch if we have to cut one before the issue is fixed upstream.

**Comment by thaJeztah** [comment]
Yup, makes sense; this would probably also give us a chance to see it fail (because of that issue).

Do we need a tracking issue for "go / no-go" before next release cut from master?

**Comment by corhere** [comment]
> Do we need a tracking issue for "go / no-go" before next release cut from master?

Yes, I think that's a good idea. The same issue could also cover "go / no-go" for #44210 once that's been reviewed and merged.

**Comment by corhere** [comment]
@thaJeztah the Go maintainers move fast. There's already a patch up for review (https://go.dev/cl/443716) which I think we can count on getting backported to Go 1.19.3 and 1.18.8.

**Comment by thaJeztah** [comment]
nice 👍

**Pull request merged**
